File: C++/0695-max-area-of-island.cpp
class Solution {
  int findPath(vector<vector<int>>& grid, int x, int y) {
    int counter = 0;
    if (grid[x][y] == 1) {
      counter++;
      grid[x][y] = 0;
      if (grid[x + 1][y] == 1) {
        counter += findPath(grid, x + 1, y);
      }
      if (grid[x - 1][y] == 1) {
        counter += findPath(grid, x - 1, y);
      }
      if (grid[x][y + 1] == 1) {
        counter += findPath(grid, x, y + 1);
      }
      if (grid[x][y - 1] == 1) {
        counter += findPath(grid, x, y - 1);
      }
    }
    return counter;
  }

 public:
  int maxAreaOfIsland(vector<vector<int>>& grid) {
    int sum = 0;
    vector<vector<int>> big_grid(grid.size() + 2,
                                 vector<int>(grid[0].size() + 2, 0));
    for (int i = 0; i < grid.size(); i++) {
      for (int j = 0; j < grid[0].size(); j++) {
        big_grid[i + 1][j + 1] = grid[i][j];
      }
    }
    for (int i = 0; i < big_grid.size(); i++) {
      for (int j = 0; j < big_grid[0].size(); j++) {
        sum = max(findPath(big_grid, i, j), sum);
      }
    }
    return sum;
  }
};

File: 0695-max-area-of-island/0695-max-area-of-island.py
# time complexity: O(n^2)
# space complexity: O(n^2)
from collections import deque
from typing import List


class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROW = len(grid)
        COL = len(grid[0])

        def traverse(r: int, c: int):
            count = 0
            queue = deque()
            queue.append((r, c))
            grid[r][c] = 0

            while queue:
                currR, currC = queue.popleft()
                count += 1

                for dR, dC in [(0, 1), (1, 0), (-1, 0), (0, -1)]:
                    nextR = currR + dR
                    nextC = currC + dC
                    if 0 <= nextR < ROW and 0 <= nextC < COL and grid[nextR][nextC] == 1:
                        grid[nextR][nextC] = 0
                        queue.append((nextR, nextC))

            return count

        result = 0
        for r in range(ROW):
            for c in range(COL):
                if grid[r][c] == 1:
                    result = max(result, traverse(r, c))

        return result


grid = [[0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0], [0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], [0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0], [
    0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0]]
print(Solution().maxAreaOfIsland(grid))
grid = [[0, 0, 0, 0, 0, 0, 0, 0]]
print(Solution().maxAreaOfIsland(grid))
